fix: updates RFC 0011 reflecting the current state of the audit scanner. (#30)
kubewarden/rfc: rfc/0011-audit-checks.md

# Detailed design

The audit check will run at specified intervals and will produce audit reports using the `PolicyReport` format.
There is going to be one `PolicyReport` object per namespaced resource. As well as a `ClusterPolicyReport` 
for each cluster wide resource 

> Details about the `PolicyReport` format can be found inside of this [RFC](https://github.com/kubewarden/rfc/blob/main/rfc/0012-policy-report.md).

Reviewer: ```suggestion
Details about the `PolicyReport` format can be found in this [RFC](https://github.com/kubewarden/rfc/blob/main/rfc/0012-policy-report.md).
```

If a policy makes use of username/groups at evaluation time, this policy will not
produce meaningful evaluation results during the audit checks.

The policy author should make this clear by adding the following field

> inside of the `metadata.yml` file of the policy:

Reviewer: ```suggestion
in the `metadata.yml` file of the policy:
```

cluster level. The initial step involves collecting these policies and
categorizing them based on the types of cluster-wide resources they examine.
The `GroupVersionResource` identifier helps in recognizing the specific
Kubernetes resource types involved.

**Note**: Only policies that review CREATE events are taken into account.

#### Step 2: fetch Kubernetes resources to be audited

After determining the policies for each resource type, the next task is to
gather the instances of these resources present in the cluster. This is done by
querying the API server, often using pagination to manage the data efficiently.

Post-retrieval, the audit scanner assesses whether each resource aligns with
the criteria set out in the policies, specifically focusing on the labels
required by the policy. If a resource fits the criteria, the scanner simulates
a `CREATE` event, incorporating the audited Kubernetes object as `request.object`
within this event.

### Step 3: performing evaluations

Initially, the evaluation is executed through an HTTP request directed at the
Policy Server that enforces the policy. This step is crucial for determining

> the compliance of each resource with the established cluster policies.

Reviewer 2: I would add that we are cylcing all the resources and auditing the resources one by one against the policies they are targeted by.

By issuing the audit request against a real Policy Server, we solve these
problems:

* Policies can be hosted on registries secured with credentials. We don't have
  to worry about being able to fetch the WebAssembly modules from these
  authenticated registries
* Policies might be hosted on insecure registires/registires with custom TLS
  certificates. Again, we won't have to bother about that
* Some data used by the policy (like external OCI manifests, sigstore
  verification results) can be cached by the Policy Server and reused between
  the audit evaluations and the regular validations requests

Sending all our auditing requests to the actual Policy Server could lead to
poor performance of the regular admission requests. Currently, based on our
load testing data, one single Policy Server running with 2 workers can evaluate
about 1600 requests per second. Based on this data, we think the requests
generated by the audit checks should not clog the Policy Server. If that turns
out to be true, we can embrace different strategies to mitigate that.

Until we have real world data, it's not useful to eagerly optimize this aspect.

### Step 4: write `ClusterPolicyReport`

> The code will overwrite the old `ClusterPolicyReport` instance with a new generated

Reviewer 2: We don't overwrite but we update the old object field with new data, also I would remove "data kept in memory" as we are keeping the data just for enough time to build the report patch.

that are scoped within specific namespaces. This process is detailed in a
series of steps:

#### Step 1: find relevant policies

The initial step involves collating policies related to the namespace to be
audited. This includes:

**Excluding Namespaces**: Initially, filter out any namespaces specified by the
user as not relevant or to be ignored.
**Compiling Policies**: For the namespace being audited, compile a list of
relevant policies. This list should include:
  - All `ClusterAdmissionPolicy` resources that apply to the namespace.
  - All `AdmissionPolicy` resources that are specifically defined within the
  namespace itself.

These policies are then organized based on the Kubernetes resource types they
target. The `GroupVersionResource` identifier helps in recognizing the specific
Kubernetes resource types involved.

**Note**: Only policies that review CREATE events are taken into account.

#### Step 2: fetch Kubernetes resources to be audited

This step mirrors the process described in Step 2 for the ClusterPolicyReport.
It involves fetching the instances of Kubernetes resources present within the

> namespace from the API server, likely in a paginated manner to manage large

Reviewer 2: remove likely, we are paginating now

That's because:

* Once all the policies and involved resources are audited, the
  `PolicyReport` audit is going to be overwritten with the results of the audit

> Users can still get live data about the rejections/approvals by looking both

Reviewer: ```suggestion
Users can still get live data about rejections and approvals by looking both
```

audited resource, executing an HTTP request to the Policy Server that enforces
the policy for each one. This step is crucial for determining the compliance of
each resource with the established cluster policies.

We will extend Policy Server to have a new endpoint for the audit checks. While
the Kubernetes API server uses `/validate/POL_ID`, the audit scanner will file
its requests against `/audit/POL_ID` endpoint.

The `/audit` endpoint will be different from the `/validate` one in these
regards:

* Generate different Prometheus metrics: we don't want the overall `/validate`
  metrics to be polluted by the audit checks
* Generate different Open Telemetry traces: again, these should not be mixed
  with the `/validate` ones. We might even want to turn them off to make the
  audit execution faster
* Policies in `monitor` mode should still behave like the ones in `protect`
  mode. They have to produce rejections if needed

By issuing the audit request against a real Policy Server, we solve these
problems:

* Policies can be hosted on registries secured with credentials. We don't have
  to worry about being able to fetch the WebAssembly modules from these
  authenticated registries

> * Policies might be hosted on insecure registires/registires with custom TLS

Reviewer: ```suggestion
* Policies might be hosted on insecure registries or registries with custom TLS
```

  certificates. Again, we won't have to bother about that
* Some data used by the policy (like external OCI manifests, sigstore
  verification results) can be cached by the Policy Server and reused between
  the audit evaluations and the regular validations requests

Sending all our auditing requests to the actual Policy Server could lead to
poor performance of the regular admission requests. Currently, based on our
load testing data, one single Policy Server running with 2 workers can evaluate
about 1600 requests per second. Based on this data, we think the requests
generated by the audit checks should not clog the Policy Server. If that turns

> out to be true, we can embrace different strategies to mitigate that.

Reviewer: ```suggestion
out to be false, we can embrace different strategies to mitigate that.
```

the policies, through simulated events or direct policy checks.

#### Step 4: write `PolicyReport`

The code updates the existing `PolicyReport` instance by patching it with
the newly generated information. This approach ensures that the report remains
current without the need to synchronize previous contents with the latest
updates.

## Changes to Policy Server

A new rest API endpoint is going to be created called `/audit`.

The endpoint will receive POST requests similar to the ones generated by the
Kubernetes API server.

All the policies hosted by the Policy Server will be exposed via `/audit/POL_ID`
endpoints.

The `/audit` API will behave in a slightly different way compared to the current `/validate`
endpoint. The differences have been illustrated by the previous section of things
RFC.

## Changes to ClusterAdmissionPolicy and AdmissionPolicy CRDs

It's necessary a change in the policy CRDs to mark a policy to be run in the background audit checks.
The field would be called `backgroundAudit` and its default value will be `true`.
Therefore, all the policies will run in the audit checks by default, unless the user

> configure the other way. Example of policies CRDs with the `backgroundAudit` field:

Reviewer: ```suggestion
configures it the other way. Example of policies CRDs with the `backgroundAudit` field:
```

    - CHOWN
    required_drop_capabilities:
    - NET_ADMIN
```

```yaml
apiVersion: policies.kubewarden.io/v1alpha2
kind: AdmissionPolicy
metadata:
  name: psp-capabilities
spec:
  policyServer: reserved-instance-for-tenant-a
  module: registry://ghcr.io/kubewarden/policies/psp-capabilities:v0.1.3
  rules:
  - apiGroups: [""]
    apiVersions: ["v1"]
    resources: ["pods"]
    operations:
    - CREATE
    - UPDATE
  mutating: true
  backgroundAudit: true
  settings:
    allowed_capabilities:
    - CHOWN
    required_drop_capabilities:
    - NET_ADMIN
```

## Background Audit Scanner

This is a new component that implements the algorithm described before to create
all the `PolicyReport` objects and the `ClusterPolicyReport` one.

The audit scanner is a CLI program running in a non interactive way. It will be
ran via a Kubernetes CronJob every 30 minutes. In the future we can provide a
way to allow the user to configure this interval.

In the initial version, the scanner will iterate sequentially over each
namespace to generate its `PolicyReport` object for each audited resource. In
the future we might want to change this implementation to scan the namespaces
in parallel.

> By default, the scanner will look at all the namespaces defined inside of the

Reviewer: ```suggestion
By default, the scanner will look at all the namespaces defined in the
```

cluster. We might want to add a cli flag and environment variable that allows
the user to explicitly pick the namespaces to inspect. This could be used to
provide a way to our user to generate the audit report of a specified namespace
on demand, without having to wait for the next run on the scanner.

# Drawbacks

> No real time audit info are provided by this solution. Users have to rely on
> Prometheus metrics and Open Telemetry trace events to have live data coming
> from the cluster.

Reviewer: ```suggestion
No real time audit information is provided by this solution. Users have to rely on
Prometheus metrics and Open Telemetry trace events to have live data coming
from the cluster.
```
